File: ain5/spko/funktional/currying/currying.py
#!/usr/bin/python3
# Beispiel aus de.wikipedia.org/wiki/Currying

# Definition und Aufruf einer Funktion mit zwei Parametern:
def uncurried_add(x, y):
    return x + y

n = uncurried_add(3, 5)
print(n)

# das gleiche mit einer Funktion und einem Lambda mit je einem Parameter:
def curried_add(x):
    return lambda y: x + y

n = curried_add(3)(5)
print(n)

add_three = curried_add(3)
n = add_three(5)
print(n)

File: ain5/spko/funktional/currying/currying.cpp
#include <iostream>

using namespace std;

auto uncurried_add = [](int x, int y) { return x + y; };
auto curried_add = [](int x) { return [x](int y) { return x + y; }; };
auto curried_add3 = [](int x) {
  return [x](int y) { return [x, y](int z) { return x + y + z; }; };
};

int main() {
  cout << uncurried_add(3, 5) << endl;                            // 8
  cout << curried_add(3)(5) << endl;                              // 8
  cout << "curry 3 equals 10: " << curried_add3(3)(5)(2) << endl; // 10

  auto add_three = curried_add(3);
  cout << add_three(5) << endl;  // 8
  cout << add_three(12) << endl; // 15

  return 0;
}
